### DISCUSSION TITLE
How to write a png with 16 bits per channel in linear mode

### DISCUSSION THREAD

**Discussion by milasudril** [ImageMagick/ImageMagick#5543] [Help] [answered]
I am trying to "stamp" a 16 bit PNG as linear, using the following command

```
convert -verbose -set colorspace RGB __targets/test_pattern_raw.png -depth 16 -colorspace RGB PNG64:__targets/test_pattern.png
```

However, the resulting image is still encoded in sRGB:

> __targets/test_pattern_raw.png PNG 1600x1000 1600x1000+0+0 16-bit sRGB 62490B 0.040u 0:00.043
> __targets/test_pattern_raw.png=>__targets/test_pattern.png PNG 1600x1000 1600x1000+0+0 16-bit TrueColorAlpha **sRGB** 78794B 0.300u 0:00.184

Also
```
identify __targets/test_pattern.png
```
>__targets/test_pattern.png PNG 1600x1000 1600x1000+0+0 16-bit sRGB 78794B 0.000u 0:00.000

Notice that it appears to work if I only use 8 bits:

`convert -verbose -set colorspace RGB __targets/test_pattern_raw.png -depth 16 -colorspace RGB PNG:__targets/test_pattern.png`

>__targets/test_pattern_raw.png PNG 1600x1000 1600x1000+0+0 16-bit sRGB 62490B 0.040u 0:00.042
>__targets/test_pattern_raw.png=>__targets/test_pattern.png PNG 1600x1000 1600x1000+0+0 8-bit **RGB** 41647B 0.160u 0:00.101

**Comment by fmw42** [comment]
`convert lena.png -colorspace RGB lena_rgb.png`

Note below the colorspace is RGB and the Gamma is 1

```
 identify -verbose lena_rgb.png
Image:
  Filename: lena_rgb.png
  Format: PNG (Portable Network Graphics)
  Mime type: image/png
  Class: DirectClass
  Geometry: 256x256+0+0
  Units: Undefined
  Colorspace: RGB    <----
  Type: TrueColor
  Base type: Undefined
  Endianness: Undefined
  Depth: 8-bit
  Channel depth:
    red: 8-bit
    green: 8-bit
    blue: 8-bit
  Channel statistics:
    Pixels: 65536
    Red:
      min: 9  (0.0352941)
      max: 255 (1)
      mean: 128.274 (0.503034)
      standard deviation: 63.6492 (0.249605)
      kurtosis: -1.15426
      skewness: -0.283565
      entropy: 0.957102
    Green:
      min: 0  (0)
      max: 250 (0.980392)
      mean: 42.9051 (0.168255)
      standard deviation: 41.8275 (0.164029)
      kurtosis: 1.03528
      skewness: 1.289
      entropy: 0.902673
    Blue:
      min: 6  (0.0235294)
      max: 175 (0.686275)
      mean: 40.6504 (0.159413)
      standard deviation: 28.7189 (0.112623)
      kurtosis: 1.78129
      skewness: 1.44184
      entropy: 0.888312
  Image statistics:
    Overall:
      min: 0  (0)
      max: 255 (1)
      mean: 70.6097 (0.276901)
      standard deviation: 44.7318 (0.175419)
      kurtosis: -0.326088
      skewness: 0.941119
      entropy: 0.916029
  Rendering intent: Undefined
  Gamma: 1  <----
  Background color: rgb(255,255,255)
  Border color: rgb(223,223,223)
  Matte color: rgb(189,189,189)
  Transparent color: rgb(0,0,0)
  Interlace: None
  Intensity: Rec709Luminance
  Compose: Over
  Page geometry: 256x256+0+0
  Dispose: Undefined
  Iterations: 0
  Compression: Zip
  Orientation: Undefined
  Properties:
    date:create: 2022-09-09T15:49:49+00:00
    date:modify: 2022-09-09T15:49:49+00:00
    date:timestamp: 2022-09-09T15:50:08+00:00
    png:bKGD: chunk was found (see Background color, above)
    png:gAMA: gamma=1 (See Gamma, above)
    png:IHDR.bit-depth-orig: 8
    png:IHDR.bit_depth: 8
    png:IHDR.color-type-orig: 2
    png:IHDR.color_type: 2 (Truecolor)
    png:IHDR.interlace_method: 0 (Not interlaced)
    png:IHDR.width,height: 256, 256
    png:text: 4 tEXt/zTXt/iTXt chunks were found
    png:tIME: 2022-09-09T15:49:49Z
    signature: e63f4d4c52b5130097d7aa27bbd5013248277c239193fe6a80ddf601442abbbd
    Software: GraphicConverter
  Artifacts:
    filename: lena_rgb.png
    verbose: true
  Tainted: False
  Filesize: 113399B
  Number pixels: 65536
  Pixels per second: 16.1523MB
  User time: 0.000u
  Elapsed time: 0:01.004
  Version: ImageMagick 6.9.12-62 Q16 aarch64 17430 https://legacy.imagemagick.org

```

**Reply by milasudril** [reply]
Notice that there is no problem if I save the data in 8 bit, as is shown by this answer. The problem appears when I choose PNG64.

**Comment by snibgo** [comment]
@milasudril: What version of IM, on what platform? Please always state these when asking questions.

```
magick toes.png -set colorspace RGB out.png

magick toes.png -set colorspace RGB PNG64:out.png
```
These work fine for me, with IM v7.1.0-42 on Windows 8.1. EDIT: It also works with IM v6.9.9-50. I mean that gamma is set to 1.0 in the output PNG.

**Reply by milasudril** [reply]
convert --version
Version: ImageMagick 6.9.11-60 Q16 x86_64 2021-01-25 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP(4.5) 
Delegates (built-in): bzlib djvu fftw fontconfig freetype heic jbig jng jp2 jpeg lcms lqr ltdl lzma openexr pangocairo png tiff webp wmf x xml zlib

On Linux Mint 21 Vanessa

This is the full script:

```
git clone https://github.com/milasudril/test_pattern
cd test_pattern
inkscape -o __targets/test_pattern_raw.png -d 192 --export-png-color-mode=RGBA_16 test_pattern.svg
convert -verbose -set colorspace RGB __targets/test_pattern_raw.png -depth 16 -colorspace RGB PNG64:__targets/test_pattern.png
identify __targets/test_pattern.png
```
I have similar problems with the .rgba recipe, but it is of cause harder to verify.

**Reply by milasudril** [reply]
Anyway, here is the input image
![test_pattern_raw](https://user-images.githubusercontent.com/734628/189473499-2bb4b9ec-ef7c-40c1-9b6a-4367a97ab3b9.png)
As exported from inkscape. With it you can also verify that pixels values matches the image metadata, since it includes gamma test pattern in form of stripes next to a gradient. The -3 dB mark should align with the point where the gradient has the same intensity as the averaged stripes.

**Comment by urban-warrior** [comment]
Thanks for the problem report.  We can reproduce it and will have a patch to fix it in the GIT main branch @ https://github.com/ImageMagick/ImageMagick later today.  The patch will be available in the beta releases of ImageMagick @ https://imagemagick.org/archive/beta/ by sometime tomorrow.

**Reply by milasudril** [reply]
I guess I should report a separate issue for the problem with .rgba? The behaviour is essentially the same, so the fix could potentially fix this issue as well:

```
if(input_color_space != output_color_space)
{
    convert_color_space();
}
```

is of cause the correct solution, but maybe the choice of output color space is affected by the output format. I that case IM should also issue a warning.

**Comment by urban-warrior** [comment, marked as answer]
Download the latest beta release of ImageMagick and try your command-line.  If it continues to fail for PNG, RGBA, etc., post a use case here we can reproduce and we will investigate further.

**Reply by milasudril** [reply]
I can confirm that both rgba and png64 works with the fix.
Here is a comparison rendered with `glEnable(GL_FRAMEBUFFER_SRGB)`

Before:
![before_fix](https://user-images.githubusercontent.com/734628/189497661-eb0f91ed-d8c1-4f3f-bb02-56622ebfb8f3.png)

After:
![after_fix](https://user-images.githubusercontent.com/734628/189497950-3d51e9cb-3eb6-4156-8915-c485701702b8.png)
